src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +65,707 @@ namespace Microsoft.CodeAnalysis.FindSymbols
+                shouldContinueSearching: IsNonSealedClass,
+                transitive: transitive,
+                cancellationToken: cancellationToken);
+        }
+
+        private static async Task<IEnumerable<INamedTypeSymbol>> FindTypesInProjectAsync(
+            HashSet<INamedTypeSymbol> sourceAndMetadataTypes,
+            Project project,
+            SearchDocumentAsync findImmediatelyDerivedTypesInDocumentAsync,
+            Func<INamedTypeSymbol, bool> shouldContinueSearching,
+            bool transitive,
+            CancellationToken cancellationToken)
+        {
+            // We're going to be sweeping over this project over and over until we reach a 
+            // fixed point.  In order to limit GC and excess work, we cache all the sematic
+            // models and DeclaredSymbolInfo for hte documents we look at.
+            // Because we're only processing a project at a time, this is not an issue.
+            var cachedModels = new ConcurrentSet<SemanticModel>();
+            var cachedInfos = new ConcurrentSet<IDeclarationInfo>();
+
+            var finalResult = new HashSet<INamedTypeSymbol>(SymbolEquivalenceComparer.Instance);
+
+            var typesToSearchFor = new HashSet<INamedTypeSymbol>(SymbolEquivalenceComparer.Instance);
+            typesToSearchFor.AddAll(sourceAndMetadataTypes);
+
+            var typeNamesToSearchFor = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            var derivesFromSystemObject = sourceAndMetadataTypes.Any(t => t.SpecialType == SpecialType.System_Object);
+            var derivesFromSystemEnum = sourceAndMetadataTypes.Any(t => t.SpecialType == SpecialType.System_Enum);
+            var derivesFromSystemValueType = sourceAndMetadataTypes.Any(t => t.SpecialType == SpecialType.System_ValueType);
+            var derivesFromSystemMulticastDelegate = sourceAndMetadataTypes.Any(t => t.SpecialType == SpecialType.System_MulticastDelegate);
+
+            var inheritanceInfo = new InheritanceInfo(
+                derivesFromSystemObject, 
+                derivesFromSystemValueType,
+                derivesFromSystemEnum,
+                derivesFromSystemMulticastDelegate,
+                typeNamesToSearchFor);
+
+            while (typesToSearchFor.Count > 0)
             {
-                // 1) Compare searchSymbol and symbolToMatch using SymbolEquivalenceComparer.IgnoreAssembliesInstance
-                if (!SymbolEquivalenceComparer.IgnoreAssembliesInstance.Equals(searchSymbol, symbolToMatch, equivalentTypesWithDifferingAssemblies.Object))
-                {
-                    // 2) If the symbols are NOT equivalent ignoring assemblies, then they cannot be equivalent.
-                    return false;
-                }
+                // Only bother searching for derived types of non-sealed classes we're passed in.
+                typeNamesToSearchFor.AddRange(typesToSearchFor.Select(c => c.Name));
```

Comment:
I don't like the spooky action at a distance of this filling in something used in `inheritanceInfo`.  It seems as if this is unused.